### PR TITLE
Automation transcoder 2 1

### PR DESCRIPTION
**Pull request by RyanRHall**

### CODE REVIEW

File: contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicA2_1.sol [hunk 1]
```diff
@@ -417,6 +414,9 @@ contract KeeperRegistryLogicA2_1 is
       bytes[] memory offchainConfigs
     ) = abi.decode(encodedUpkeeps, (uint256[], Upkeep[], address[], bytes[], bytes[], bytes[]));
     for (uint256 idx = 0; idx < ids.length; idx++) {
+      if (address(upkeeps[idx].forwarder) == ZERO_ADDRESS) {
+        upkeeps[idx].forwarder = new AutomationForwarder(ids[idx], upkeeps[idx].target, address(this));
+      }
```
Comment (ogtownsend) [inline contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicA2_1.sol:419]: Was the forwarder not assigned in v2_0 then, or was it assigned elsewhere?

File: contracts/src/v0.8/automation/UpkeepFormat.sol [hunk 2]
```diff
@@ -2,6 +2,10 @@
 
 pragma solidity ^0.8.0;
 
+/**
+ * @dev this struct is only maintained for backwards compatibility with MigratableKeeperRegistryInterface
+ * it should deprecated in the future in favor of MigratableKeeperRegistryInterfaceV2
```
Comment (FelixFan1992) [inline contracts/src/v0.8/automation/UpkeepFormat.sol:7]: should `be` deprecated

File: contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicA2_1.sol (hunk 1 above)
Comment (RyanRHall) [reply contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicA2_1.sol:419]: We didn't have one - it's a new feature in v2.1

**Comment by FelixFan1992** [comment]
I manually tested with
2.1: https://goerli.etherscan.io/address/0xfA1b41B247C91Bea924bD4820Ff1e4dD0FEFF864
2.0: https://goerli.etherscan.io/address/0x4944CFa9d0B7a22c73c0B1f24e625430C2984FA2#readContract
1.2: https://goerli.etherscan.io/address/0x0ee0d15d6f2b20739e498447ca60a3d612c45c70#writeContract
transcoder: https://goerli.etherscan.io/address/0x6b51D23ecEed3Bc1B9E24d8C7fCD4677E8594792#code

**Review by FelixFan1992 — APPROVED** [review]

**Pull request merged**